node/libs/concurrency/src/sync/prunable_queue.rs

```diff
@@ -0,0 +1,148 @@
+//! Prunable queue mimics a producer-consumer channel, in enabling
+//! waiting on the receiver side for new items, while providing pruning functionality.
+//! The pruning takes place whenever a new item is sent, based on a specified predicate.
+//!
+//! The queue facilitates the asynchronous return of processing result for each added item.
+//!
+//! The separation of `Sender` and `Receiver` is employed primarily because `Receiver` requires
+//! a mutable reference to the signaling channel, unlike `Sender`, hence making it undesirable to
+//! be used in conjunction.
+
+use crate::{
+    ctx::{self, Ctx},
+    oneshot,
+    sync::{self, watch},
+};
+use std::{collections::VecDeque, sync::Arc};
+use tokio::sync::Mutex;
+
+pub fn new<T, U>(
+    pruning_predicate: Box<dyn Sync + Send + Fn(&T, &T) -> bool>,
+) -> (Sender<T, U>, Receiver<T, U>) {
+    let queue: Mutex<VecDeque<(T, oneshot::Sender<U>)>> = Mutex::new(VecDeque::new());
+    // Internal signaling, to enable waiting on the receiver side for new items.
+    let (send, recv) = watch::channel(false);
+
+    let shared = Arc::new(Shared {
+        queue,
+        has_items_send: send,
+    });
+
+    let sender = Sender {
+        shared: shared.clone(),
+        pruning_predicate: pruning_predicate,
+    };
+
+    let receiver = Receiver {
+        shared: shared.clone(),
+        has_items_recv: recv,
+    };
+
+    return (sender, receiver);
+}
+
+pub struct Shared<T, U> {
+    queue: Mutex<VecDeque<(T, oneshot::Sender<U>)>>,
```

Comment:
synchronous mutex should be good enough here (or watch::Sender, as mentioned in another comment). It will also make the enqueue sync.

Reply:
Doesn't sync `Mutex` halt async execution environment?

Reply:
Doesn't sync `Mutex` halts async execution environments?

Reply:
Sync mutex blocks the thread, so it is one thread less for other tasks to use. However same applies to just running a task on a thread which takes nontrivial amount of time. Fast operations are ok to be executed under a blocking mutex on a runtime. No deadlock is possible, because you cannot hold a lock when a task yields (MutexGuard is !Send).

Reply:
Fixed, changes here: https://github.com/matter-labs/era-consensus/pull/56/commits/84a919cf0c41ca900b4c5a5f8bebba1d733d0133

Reply:
Fixed, changes here: https://github.com/matter-labs/era-consensus/pull/56/commits/6d02da048b899ed944966d2c69d8b8230e036121

Reply:
Fixed, changes here: https://github.com/matter-labs/era-consensus/pull/56/commits/ba22e7400b699d7daf847bea570c36f65f9cd69f